onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp

```diff
@@ -0,0 +1,515 @@
+/*++
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+
+Licensed under the MIT License.
+
+Module Name:
+
+    sqnbitgemm_kernel_neon.h
+
+Abstract:
+
+    This module implements the float/quantized n-bit integer matrix
+    multiplication kernels for ARM NEON.
+
+--*/
+
+#include "sqnbitgemm.h"
+
+#include <arm_neon.h>
+
+#include <algorithm>
+#include <array>
+#include <utility>
+
+//
+// Hardware-specific kernel type.
+//
+struct MLAS_SQNBIT_GEMM_KERNEL_NEON {
+};
+
+namespace
+{
+
+template <typename IterationFn, size_t... Indices>
+MLAS_FORCEINLINE void
+UnrolledLoopIterations(IterationFn&& f, std::index_sequence<Indices...> /* indices */)
+{
+    (f(Indices), ...);
+}
+
+template <size_t N, typename IterationFn>
+MLAS_FORCEINLINE void
+UnrolledLoop(IterationFn&& f)
+{
+    UnrolledLoopIterations(std::forward<IterationFn>(f), std::make_index_sequence<N>());
+}
+
+MLAS_FORCEINLINE float32x4_t
+FoldAccumulators(float32x4_t a0, float32x4_t a1, float32x4_t a2, float32x4_t a3)
+{
+    // aN: aN_0 aN_1 aN_2 aN_3
+
+    float32x4_t b0 = vzip1q_f32(a0, a1);  // a0_0 a1_0 a0_1 a1_1
+    float32x4_t b1 = vzip2q_f32(a0, a1);  // a0_2 a1_2 a0_3 a1_3
+    float32x4_t b2 = vzip1q_f32(a2, a3);  // a2_0 a3_0 a2_1 a3_1
+    float32x4_t b3 = vzip2q_f32(a2, a3);  // a2_2 a3_2 a2_3 a3_3
+
+    // a0_0 a1_0 a2_0 a3_0
+    a0 = vreinterpretq_f32_f64(vzip1q_f64(vreinterpretq_f64_f32(b0), vreinterpretq_f64_f32(b2)));
+    // a0_1 a1_1 a2_1 a3_1
+    a1 = vreinterpretq_f32_f64(vzip2q_f64(vreinterpretq_f64_f32(b0), vreinterpretq_f64_f32(b2)));
+    // a0_2 a1_2 a3_2 a3_2
+    a2 = vreinterpretq_f32_f64(vzip1q_f64(vreinterpretq_f64_f32(b1), vreinterpretq_f64_f32(b3)));
+    // a0_3 a1_3 a2_3 a3_3
+    a3 = vreinterpretq_f32_f64(vzip2q_f64(vreinterpretq_f64_f32(b1), vreinterpretq_f64_f32(b3)));
+
+    return vaddq_f32(vaddq_f32(a0, a1), vaddq_f32(a2, a3));
+}
+
+template <size_t Capacity>
+MLAS_FORCEINLINE std::array<float32x4_t, Capacity / 4>
+LoadDataAndZeroPad(const float* data, size_t count)
+{
+    static_assert(Capacity % 4 == 0, "Capacity must be divisible by 4.");
+
+    count = std::min(count, Capacity);
+
+    std::array<float32x4_t, Capacity / 4> dst{};
+
+    size_t vi = 0;  // vector index
+
+    // handle 4 values at a time
+    while (count > 3) {
+        dst[vi] = vld1q_f32(data);
+
+        vi += 1;
+        data += 4;
+        count -= 4;
+    }
+
+    // handle remaining values
+    if (count > 0) {
+        dst[vi] = vsetq_lane_f32(data[0], dst[vi], 0);
+    }
+
+    if (count > 1) {
+        dst[vi] = vsetq_lane_f32(data[1], dst[vi], 1);
+    }
+
+    if (count > 2) {
+        dst[vi] = vsetq_lane_f32(data[2], dst[vi], 2);
+    }
```

Comment:
the overhead is relatively too big. The functionality is to load 4 floatx4 vector, which only needs 4 instructions, but the while loop, 3 ifs and the std::min and std::array cost much more than 4 instructions.

Reply:
yeah, using std::array looks nice, but it uses extra memory buffer, which is not necessary most of the time.

Reply:
this did not show up in my measurements. perhaps we can revisit it later if it becomes significant.

see this comment:
https://github.com/microsoft/onnxruntime/pull/18031/files/04cb16143dc3cc44e812e21cc17e08a88146431e#r1393487945